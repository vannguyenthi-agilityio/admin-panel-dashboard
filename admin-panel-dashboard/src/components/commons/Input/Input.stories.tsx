import type { Meta, StoryObj } from '@storybook/react';
import Input from '.';

import {
  MailIcon,
} from '@/components/Icons';

const meta: Meta<typeof Input> = {
  args: {
    disabled: false,
    id: 'input-field',
    name: 'input-field',
    label: 'Input Field',
    type: 'text',
  },
  argTypes: {
    'aria-invalid': {
      control: { type: 'boolean' },
    },
    borderHide: {
      type: 'boolean',
    },
    defaultValue: {
      type: 'string',
    },
    disabled: {
      type: 'boolean',
    },
    leftElement: {
      table: { disable: true },
    },
    rightElement: {
      table: { disable: true },
    },
    placeholder: {
      type: 'string',
    },
    type: {
      control: { type: 'select' },
      options: ['text', 'email', 'password'],
    },
  },
  component: Input,
  tags: ['autodocs'],
  title: 'Components/Commons/Input',
};

export default meta;
type Story = StoryObj<typeof Input>;

export const WithLabel: Story = {};

export const WithoutLabel: Story = {
  args: {
    label: undefined,
  },
};

export const WithoutBorder: Story = {
  args: {
    borderHide: true,
  },
};

export const Text: Story = {
  args: {
    defaultValue: 'Text',
    type: 'text',
  },
};

export const Email: Story = {
  args: {
    defaultValue: 'example@site.com',
    type: 'email',
    required: true
  },
};

export const Password: Story = {
  args: {
    defaultValue: 'password',
    type: 'password'
  },
};

export const Disabled: Story = {
  args: {
    defaultValue: undefined,
    disabled: true,
  },
};

export const WithPlaceholder: Story = {
  args: {
    label: undefined,
    placeholder: 'Search...',
    leftElement: <MailIcon />,
  },
};

export const ErrorWithValue: Story = {
  args: {
    'aria-invalid': true,
    defaultValue: 'Value',
  },
};

export const ErrorWithoutValue: Story = {
  args: {
    'aria-invalid': true,
  },
};

export const WithLeftElementAndLabel: Story = {
  args: {
    leftElement: <MailIcon className="h-4 w-4" />,
  },
};

export const WithLeftElementAndWithoutLabel: Story = {
  args: {
    label: undefined,
    leftElement: <MailIcon className="h-4 w-4" />,
    placeholder: 'Email',
  },
};

export const WithRightElementAndLabel: Story = {
  args: {
    rightElement: <MailIcon className="h-4 w-4" />,
  },
};

export const WithRightElementAndWithoutLabel: Story = {
  args: {
    label: undefined,
    rightElement: <MailIcon className="h-4 w-4" />,
    placeholder: 'Email',
  },
};

export const ErrorInput: Story = {
  args: {
    type: 'text',
    name: 'Error input',
    errorMessage: 'Error message',
  },
};

