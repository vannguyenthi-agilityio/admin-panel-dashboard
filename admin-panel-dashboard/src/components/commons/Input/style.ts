import { cva } from 'class-variance-authority';

export const inputStyles = cva(
  "peer block w-full flex-1 font-primary text-sm font-light appearance-none rounded-lg border-gray-light bg-white px-4 text-sm text-tertiary placeholder-text-gray focus:outline-none hover:enabled:border-dark-gray aria-invalid:!border-red disabled:text-gray disabled:cursor-not-allowed",
  {
    defaultVariants: {
      hasBorder: true,
      hasLabel: false,
      hasLeftElement: false,
      hasRightElement: false
    },
    variants: {
      hasBorder: {
        true: "border",
        false: "focus:outline-darker h-[33px]"
      },
      hasLabel: {
        false: "",
        true: "placeholder:opacity-0 focus:placeholder:opacity-100"
      },
      hasLeftElement: {
        false: "",
        true: "pl-9"
      },
      hasRightElement: {
        false: "",
        true: "pr-9"
      },
      size: {
        sm: "py-1",
        default: "py-[9px]"
      }
    }
  }
);
