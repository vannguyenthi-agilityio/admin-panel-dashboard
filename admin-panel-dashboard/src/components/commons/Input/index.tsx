import { memo, forwardRef } from "react";

// Types
import { IInput } from "@/types";

// Helpers
import { clsxMerge } from '@/helpers';

import { inputStyles } from './style';

const Input = forwardRef<HTMLInputElement, IInput>((
  {
    borderHide,
    className,
    label,
    required,
    leftElement,
    rightElement,
    placeholder,
    errorMessage,
    size = "default",
    ...props
  },
  ref
) => {
  const classNameLabel = required
    ? "text-sm font-medium after:content-['*'] after:text-red-500 after:text-md"
    : "text-sm font-medium";

  return (
    <div className="w-full">
      {label && (
        <label htmlFor={props.id} className={classNameLabel}>
          {label}
        </label>
      )}
      <div className="relative w-full mt-1">
        {leftElement && (
          <div className="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-4 text-dark-gray">
            {leftElement}
          </div>
        )}
        <input
          className={clsxMerge(
            inputStyles({
              hasBorder: !borderHide,
              hasLabel: !!label,
              hasLeftElement: !!leftElement,
              size
            }),
            className
          )}
          placeholder={placeholder ?? " "}
          ref={ref}
          {...props}
        />

        {rightElement && (
          <div className="absolute inset-y-0 right-0 flex items-center pr-4 text-dark-gray">
            {rightElement}
          </div>
        )}
      </div>
      {errorMessage && (
        <p className="pt-1 text-xs text-red-500 dark:text-red-500">
          {errorMessage}
        </p>
      )}
    </div>
  );
});

Input.displayName = "Input";
export default memo(Input);
