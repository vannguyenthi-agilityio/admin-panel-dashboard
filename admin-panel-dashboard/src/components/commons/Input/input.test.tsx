import {
  render,
  screen,
  fireEvent
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Input from '.'

describe('Input component', () => {
  it('Renders input component', () => {
    const {container} = render(<Input
      name='Input field'
      type='text'/>
    )
    expect(container).toMatchSnapshot();
  });

  it('Handles defaultValue and onChange event', () => {
    const handleChange = jest.fn();
    render(<Input
      defaultValue='Initial value'
      onChange={handleChange}
      name='Input field'
      label='Input field'
      id="Input field"
      type='text'/>
    )

    const inputElement = screen.getByLabelText('Input field');
    expect(inputElement).toHaveValue('Initial value');

    // Simulate a change event
    fireEvent.change(inputElement, { target: { value: 'New value' } });
    expect(handleChange).toHaveBeenCalled();
  });

  it('Should show error message when have error message', () => {
    render(<Input
      errorMessage='This is an error'
      name='Input field'
      type='text'
    />);
    const errorMessage = screen.getByText('This is an error');
    expect(errorMessage).toBeInTheDocument();
  })

  it('Should not render error message when have not error message', () => {
    render(<Input
      name='Input field'
      type='text'
    />);
    const errorMessage = screen.queryByText('This is an error');
    expect(errorMessage).not.toBeInTheDocument();
  })
})
