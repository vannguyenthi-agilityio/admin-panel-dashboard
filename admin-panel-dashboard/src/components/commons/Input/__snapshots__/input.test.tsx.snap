// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Renders input component 1`] = `
<div>
  <div
    class="w-full"
  >
    <div
      class="relative w-full mt-1"
    >
      <input
        class="peer block w-full flex-1 font-primary font-light appearance-none rounded-lg border-gray-light bg-white px-4 text-sm text-tertiary placeholder-text-gray focus:outline-none hover:enabled:border-dark-gray aria-invalid:!border-red disabled:text-gray disabled:cursor-not-allowed border py-[9px]"
        name="Input field"
        placeholder=" "
        type="text"
      />
    </div>
  </div>
</div>
`;
